File: ZezeJavaTest/src/UnitTest/Zeze/Net/TestRpc.java
package UnitTest.Zeze.Net;


import Zeze.Net.*;
import Zeze.Transaction.*;
import Zeze.Util.Factory;
import Zeze.Util.ManualResetEvent;
import junit.framework.TestCase;

import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;

public class TestRpc extends TestCase{
	
	 ManualResetEvent connected = new ManualResetEvent(false);
	
	public final void testRpcSimple() throws InterruptedException {
		Service server = new Service("TestRpc.Server");
		Zeze.Util.Task.tryInitThreadPool(null, null, null);
		FirstRpc forid = new FirstRpc();
		Factory<Protocol> f =  () -> new FirstRpc();
		server.AddFactoryHandle(forid.getTypeId(), new Service.ProtocolFactoryHandle(f,x-> ProcessFirstRpcRequest(x)));

		AsyncSocket servetrSocket = server.NewServerSocket("127.0.0.1", 5000);
		Client client = new Client(this);
		client.AddFactoryHandle(forid.getTypeId(), new Service.ProtocolFactoryHandle(() -> new FirstRpc()));

		AsyncSocket clientSocket = client.NewClientSocket("127.0.0.1", 5000, null);
		connected.WaitOne();

		FirstRpc first = new FirstRpc();
		first.Argument.setInt1(1234);
		//Console.WriteLine("SendFirstRpcRequest");
		first.SendForWait(clientSocket).Wait();
		//Console.WriteLine("FirstRpc Wait End");
		assert first.Argument.getInt1() == first.Result.getInt1();
	}

	

	public final int ProcessFirstRpcRequest(Protocol p) {
		FirstRpc rpc = (FirstRpc) p;
		rpc.Result.Assign(rpc.Argument);
		rpc.SendResult();
		System.out.println("ProcessFirstRpcRequest result.Int1=" + rpc.Result.getInt1());
		return Procedure.Success;
	}

	public static class FirstRpc extends Rpc<demo.Module1.Value, demo.Module1.Value> {
		public FirstRpc() {
			Argument = new demo.Module1.Value();
			Result = new demo.Module1.Value();
		}
		
		@Override
		public int getModuleId() {
			return 1;
		}

		@Override
		public int getProtocolId() {
			return 1;
		}
	}
	public static class Client extends Service {
		private TestRpc test;
		public Client(TestRpc test) {
			super("TestRpc.Client");
			this.test = test;
		}
		@Override
		public void OnSocketConnected(AsyncSocket so) {
			super.OnSocketConnected(so);
			test.connected.Set();
		}
	}
}

File: UnitTest/Zeze/Net/TestRpc.cs
using System;
using System.Collections.Generic;
using System.Text;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Zeze.Net;
using System.Net;
using Zeze.Transaction;
using System.Threading;
using System.Threading.Tasks;
using Zeze.Util;

namespace UnitTest.Zeze.Net
{
    [TestClass]
    public class TestRpc
    {
        [TestMethod]
        public void TestRpcSimple()
        {
            var server = new Service("TestRpc.Server");

            var forid = new FirstRpc();
            server.AddFactoryHandle(forid.TypeId, new Service.ProtocolFactoryHandle()
            {
                Factory = () => new FirstRpc(),
                Handle = ProcessFirstRpcRequest,
            });

            var servetrSocket = server.NewServerSocket(IPAddress.Any, 5000, null);
            var client = new Client(this);
            client.AddFactoryHandle(forid.TypeId, new Service.ProtocolFactoryHandle()
            {
                Factory = () => new FirstRpc(),
            });

            var clientSocket = client.NewClientSocket("127.0.0.1", 5000, null, null);
            connected.WaitOne();

            var first = new FirstRpc();
            first.Argument.Int1 = 1234;
            //Console.WriteLine("SendFirstRpcRequest");
            first.SendAsync(clientSocket).Wait();
            //Console.WriteLine("FirstRpc Wait End");
            Assert.AreEqual(first.Argument.Int1, first.Result.Int1);
        }

        readonly ManualResetEvent connected = new(false);

        public Task<long> ProcessFirstRpcRequest(Protocol p)
        {
            var rpc = p as FirstRpc;
            rpc.Result.Assign(rpc.Argument);
            rpc.SendResult();
            Console.WriteLine("ProcessFirstRpcRequest result.Int1=" +  rpc.Result.Int1);
            return Task.FromResult(ResultCode.Success);
        }

        public class FirstRpc : Rpc<demo.Module1.Value, demo.Module1.Value>
        {
            public override int ModuleId => 1;

            public override int ProtocolId => 1;
        }
        public class Client : Service
        {
            readonly TestRpc test;
            public Client(TestRpc test) : base("TestRpc.Client")
            {
                this.test = test;
            }
            public override void OnSocketConnected(AsyncSocket so)
            {
                base.OnSocketConnected(so);
                test.connected.Set();
            }
        }
    }
}
